### ISSUE TITLE
inconsistant result between python and c++ conv2d operation

### ISSUE THREAD

**Issue by libine** [clab/dynet#536]
Hi  Guys:

    Tesing the conv2d operation using the same input and same kernel,   One inconsistant result was found which is as follows:

C++ Code:

struct TestModel {

  Parameter param_kernel;
  explicit TestModel(Model& model) {
    param_kernel = model.add_parameters({2, 2, 2, 3});
  }

  void Eval(ComputationGraph& cg) {

    std::vector<float> param_kernel_vals = {.011f, .022f, .033f, .012f, .022f, .032f, .013f, .023f, .033f,
                                                                    .111f, -.122f, -.033f, -.112f, -.022f, -.132f, -.113f, -.123f, -.133f,
                                                                    .211f, .222f, .233f, .212f, .222f, .232f};
    TensorTools::set_elements(param_kernel.get()->values, param_kernel_vals);

    std::vector<float> conv2d_batch_vals(2 * 3 * 2 * 1);
    for (unsigned i = 0; i < conv2d_batch_vals.size(); ++i) {
      conv2d_batch_vals[i] = i * 0.011f + (i+1) * 0.001f;
    }

    Expression x = input(cg, Dim({2, 3, 2}, 1), conv2d_batch_vals);
    cout << x.dim() << endl;
    Expression x_shape = reshape(x, Dim({2, 6}, 1));
    cout << x_shape.value() << endl << endl;

    Expression kernel = parameter(cg, param_kernel);
    cout << kernel.dim() << endl;
    Expression kernel_shape = reshape(kernel, Dim({2, 12}, 1));
    cout << kernel_shape.value() << endl << endl;

    vector<unsigned> stride = {1, 1}; bool is_valid = false;
    Expression y = conv2d(x, kernel, stride, is_valid);
    cout << y.dim() << endl << endl;
    Expression y_shape = reshape(y, Dim({2, 9}, 1));
    cout << y_shape.value() << endl << endl;
  }
};

int main(int argc, char** argv) {
  dynet::initialize(argc, argv);
  Model model;
  TestModel test_model(model);
  ComputationGraph cg;
  test_model.Eval(cg);
}

Result:
input:  Dim({2, 3, 2}, 1)
input result reshape by  Dim({2, 6}, 1)
**content:** 
0.001 0.025 0.049 0.073 0.097 0.121
0.013 0.037 0.061 0.085 0.109 0.133 

kernel:   {2, 2, 2, 3}
kernel result reshape by  Dim({2, 12}, 1)
**content:**
0.011  0.033  0.022  0.013  0.033 -0.122 -0.112 -0.132 -0.123  0.211  0.233  0.222
0.022  0.012  0.032  0.023  0.111 -0.033 -0.022 -0.113 -0.133  0.222  0.212  0.232

conv2d result:  {2,3,3}
conv2d result  reshape by  Dim({2, 9}, 1)
**content:**
0.00966  0.013692  0.008799 -0.037962 -0.047322  -0.00809  0.093488  0.119312  0.042249
0.004651  0.006547  0.003597 -0.027993 -0.035985 -0.012883  0.050211  0.063243  0.023486

Python Code: 

from dynet import *
import numpy as np

kernel = np.reshape(np.loadtxt('/serving/libin/model_test/data/kernel_matrix', delimiter=','), [2, 2, 2, 3]).astype(np.float32)
input = np.reshape(np.loadtxt('/serving/libin/model_test/data/input', delimiter=' '), [2, 3, 2]).astype(np.float32)
m = Model() 
kernel_p = m.parameters_from_numpy(kernel)
input_ts = inputTensor(input)

**input_ts.value()**
array([[[ 0.001,  0.013],
        [ 0.025,  0.037],
        [ 0.049,  0.061]],

       [[ 0.073,  0.085],
        [ 0.097,  0.109],
        [ 0.121,  0.133]]])

**parameter(kernel_p).value()**
array([[[[ 0.011     ,  0.022     ,  0.033     ],
         [ 0.012     ,  0.022     ,  0.032     ]],

        [[ 0.013     ,  0.023     ,  0.033     ],
         [ 0.111     , -0.122     , -0.033     ]]],

       [[[-0.112     , -0.022     , -0.132     ],
         [-0.113     , -0.123     , -0.133     ]],

        [[ 0.211     ,  0.222     ,  0.233     ],
         [ 0.212     ,  0.222     ,  0.23199999]]]])

res = conv2d(input_ts, parameter(kernel_p), [1, 1], False)
**res.value()**
array([[[ 0.030393,  0.03004 ,  0.027001],
        [ 0.038673,  0.035896,  0.033361],
        [-0.02731 , -0.016601, -0.030092]],

       [[ 0.015183, -0.007591,  0.004733],
        [ 0.018711, -0.008911,  0.006293],
        [ 0.002927,  0.005588,  0.008249]]])
   
i  wonder why the c++ result is not the same with the python result ?   any help?   thanks a lot !

**Comment by zhisbug** [comment]
It looks like you are using different data layouts for your C++ and python code. Could you try to transpose your python numpy array?

**Comment by libine** [comment]
i  made a double-check between the python dynet code and tensorflow code,  which is as follows:

kernel = np.reshape(np.loadtxt('/serving/libin/model_test/data/kernel_matrix', delimiter=','), [2, 2, 2, 3]).astype(np.float32)
input = np.reshape(np.loadtxt('/serving/libin/model_test/data/input', delimiter=' '), [2, 3, 2]).astype(np.float32)

tf_input = tf.concat(0, np.reshape(input, (1, 2, 3, 2)))
conv_kernel = tf.Variable(kernel)
res_tf = tf.nn.conv2d(tf_input, conv_kernel, strides = [1, 1, 1, 1], padding='SAME')
init_op = tf.group(tf.initialize_all_variables(), tf.initialize_local_variables())
with tf.Session() as sess:
  sess.run(init_op)
  sess.run(tf_input)
  sess.run(conv_kernel)
  sess.run(res_tf)

array([[[[ 0.001,  0.013],
         [ 0.025,  0.037],
         [ 0.049,  0.061]],

        [[ 0.073,  0.085],
         [ 0.097,  0.109],
         [ 0.121,  0.133]]]], dtype=float32)

array([[[[ 0.011     ,  0.022     ,  0.033     ],
         [ 0.012     ,  0.022     ,  0.032     ]],

        [[ 0.013     ,  0.023     ,  0.033     ],
         [ 0.111     , -0.122     , -0.033     ]]],


       [[[-0.112     , -0.022     , -0.132     ],
         [-0.113     , -0.123     , -0.133     ]],

        [[ 0.211     ,  0.222     ,  0.233     ],
         [ 0.212     ,  0.222     ,  0.23199999]]]], dtype=float32)

array([[[[ 0.030393,  0.03004 ,  0.027001],
         [ 0.038673,  0.035896,  0.033361],
         [-0.02731 , -0.016601, -0.030092]],

        [[ 0.015183, -0.007591,  0.004733],
         [ 0.018711, -0.008911,  0.006293],
         [ 0.002927,  0.005588,  0.008249]]]], dtype=float32)

kernel file is (one row):
0.011, 0.022, 0.033, 0.012, 0.022, 0.032, 0.013, 0.023, 0.033, 0.111, -0.122, -0.033, -0.112, -0.022, -0.132, -0.113, -0.123, -0.133, 0.211, 0.222, 0.233, 0.212, 0.222, 0.232

input file is (one row):
0.001 0.013 0.025 0.037 0.049 0.061 0.073 0.085 0.097 0.109 0.121 0.133

So,  when input and kernel are same,  the python dynet result is the same with that of tensorflow 

i  made a output about the tensor's dim (c++ code) to ensure that it is consistant with that of python,  and made a output  about the content of tensor (c++ code) to see if the values are the same with that of python,  reshape operation is due to the error (when output the value of input x):

Attempted to access Tensor with more than one batch element or more than two dimensions in matrix form: {2,3,2} 

i wonder if the reshape operation will change the layout of tensor ?
how to debug or fix my problem,   any help?   thanks a lot ！

**Comment by libine** [comment]
when i delete the following code:

Expression x_shape = reshape(x, Dim({2, 6}, 1));
cout << x_shape.value() << endl << endl;

Expression kernel_shape = reshape(kernel, Dim({2, 12}, 1));
cout << kernel_shape.value() << endl << endl;

the result does not change ！

**Comment by libine** [comment]
Thanks~  @zhisbug ,    I have fixed the problem of inconsistent result of different  dl platforms.   You are right that the problem are due to the layout of tensor which is of key importance.
